ExperimentDatabase.py

```diff
@@ -75,20 +84,53 @@ def setup_measurement_items(self, items):
                                 (item[0], item[1]))
             self._conn.commit()   
 
-    def setup_cages(self):
-        pass
-
-
-
-    def add_animal(self, rfid, group_id, cage_id, remarks=''):
+    def add_animal(self, rfid, remarks=''):
         self._c.execute("INSERT INTO animal_rfid (rfid) VALUES (?)", (rfid, ))
         self._conn.commit()
+        
         animal_id = self.get_animal_id(rfid)
+        cage_id = self._get_next_cage()
+        group_id = self._get_next_group()
+
         self._c.execute("INSERT INTO animals (animal_id, group_id, cage_id, remarks, active) VALUES (?, ?, ?, ?, True)",
                         (animal_id, group_id, cage_id, remarks))
         self._conn.commit()
+        return (self.get_animal_id(rfid))
 
+    def _get_next_cage(self):
+        self._c.execute("SELECT cage_id, num_animals FROM cages WHERE full=0")
+        info = self._c.fetchone()
+        cage_id = info[0]
+        num_animals = info[1] + 1
 
+        self._c.execute("SELECT cage_max FROM experiment")
+        max = self._c.fetchone()[0]
+        
+        full = 0
+        if (num_animals == max):
+            full=1
+        self._c.execute("UPDATE cages SET num_animals=?, full=? WHERE cage_id=?", (num_animals, full, cage_id))
+        self._conn.commit()
+
+        return cage_id
+
+    def _get_next_group(self):
+        self._c.execute("SELECT group_id, num_animals FROM groups WHERE full=0")
+        info = self._c.fetchone()
+        group_id = info[0]
+        num_group_animals = info[1] + 1
+        
+        self._c.execute("SELECT num_animals, num_groups FROM experiment")
+        animals = self._c.fetchone()
+        max = animals[0] / animals[1]
```

Comment:
What happens if it's not a perfect integer? Or shall we use integer division ?

Reply:
We should ensure that it is a perfect interger when they are setting up the experiment.